src/dragonball/src/device_manager/mod.rs

```diff
@@ -0,0 +1,600 @@
+// Copyright (C) 2022 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Device manager to manage IO devices for a virtual machine.
+
+use std::io;
+use std::sync::{Arc, Mutex, MutexGuard};
+
+use arc_swap::ArcSwap;
+use dbs_address_space::AddressSpace;
+use dbs_device::device_manager::{Error as IoManagerError, IoManager, IoManagerContext};
+use dbs_device::resources::Resource;
+use dbs_device::DeviceIo;
+use dbs_interrupt::KvmIrqManager;
+use dbs_legacy_devices::ConsoleHandler;
+use dbs_utils::epoll_manager::EpollManager;
+use kvm_ioctls::VmFd;
+
+#[cfg(feature = "dbs-virtio-devices")]
+use dbs_device::resources::ResourceConstraint;
+#[cfg(feature = "dbs-virtio-devices")]
+use dbs_virtio_devices as virtio;
+#[cfg(feature = "dbs-virtio-devices")]
+use dbs_virtio_devices::{
+    mmio::{
+        MmioV2Device, DRAGONBALL_FEATURE_INTR_USED, DRAGONBALL_FEATURE_PER_QUEUE_NOTIFY,
+        DRAGONBALL_MMIO_DOORBELL_SIZE, MMIO_DEFAULT_CFG_SIZE,
+    },
+    VirtioDevice,
+};
+
+use crate::address_space_manager::GuestAddressSpaceImpl;
+use crate::error::StartMicrovmError;
+use crate::resource_manager::ResourceManager;
+use crate::vm::KernelConfigInfo;
+
+/// Virtual machine console device manager.
+pub mod console_manager;
+/// Console Manager for virtual machines console device.
+pub use self::console_manager::ConsoleManager;
+
+mod legacy;
+pub use self::legacy::{Error as LegacyDeviceError, LegacyDeviceManager};
+
+#[cfg(feature = "virtio-vsock")]
+/// Device manager for user-space vsock devices.
+pub mod vsock_dev_mgr;
+#[cfg(feature = "virtio-vsock")]
+use self::vsock_dev_mgr::VsockDeviceMgr;
+
+macro_rules! info(
+    ($l:expr, $($args:tt)+) => {
+        slog::info!($l, $($args)+; slog::o!("subsystem" => "device_manager"))
+    };
+);
+
+/// Errors related to device manager operations.
+#[derive(Debug, thiserror::Error)]
+pub enum DeviceMgrError {
+    /// Invalid operation.
+    #[error("invalid operation")]
```

Comment:
Again, if this is the message the user sees, can you add more detail ("invalid device manager operation" or similar?)